README.md
# redirect-ssl
> Connect/Express middleware to enforce https using [is-https](https://www.npmjs.com/package/is-https).

[![version][npm-v-src]][npm-v-href]
[![downloads][npm-d-src]][npm-d-href]
[![ci][ci-src]][ci-href]

## Usage

Install package:

```bash
yarn add redirect-ssl
# or
npm install redirect-ssl
```

Require and use `redirect-ssl`. Make sure to use this middlware as the first in your middleware chain (if using express see [middleware chain](http://expressjs.com/en/guide/using-middleware.html):

```js
import redirectSSL from 'redirect-ssl'
// or
const redirectSSL = require('redirect-ssl')

// Add middleware
app.use(redirectSSL)

// Using custom options
app.use(redirectSSL.create({ redirectPort: 8443 }))
```

### Disable for non-production or localhost

If you want to disable on `localhost`, use the exclude option:

```js
app.use(redirectSSL.create({
   exclude: ['localhost']
}))
```

Only enable in production environments:

```js
app.use(redirectSSL.create({
  enabled: process.env.NODE_ENV === 'production'
}))
```

## Options

### trustProxy

- Default: `true`

Trust and check `x-forwarded-proto` header for HTTPS detection.

### enabled

- Default: `true`

### redirectPort

- Default: `443`

Redirect users to this port for HTTPS. (`:443` is omitted from URL as is default for `https://` schema)

### redirectHost

- Default: `req.headers.host`

Redirects using this value as host, if omitted will use request host for redirects.

**NOTE** It should not contain schema or trailing slashes. (Example: `google.com`)

### redirectUnknown

- Default: `true`

Redirect when no SSL detection method is available too. **disable** this option if you encounter **redirect loops**.

### statusCode

- Default: `307` *Temporary Redirect*

Status code when redirecting. The reason of choosing `307` for default is:
- It prevents changing method from `POST` TO `GET` by user agents. (If you don't care, use `302` *Found*)
- Is temporary so if for any reason HTTPS disables on server clients won't hurt. (If you need permanent, use `308` *Permanent Redirect* or `301` *Moved Permanently*)
- See [This question](https://stackoverflow.com/questions/42136829/whats-difference-between-http-301-and-308-status-codes), [307 on MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/307), and [RFC 7231 section 6.4.7](https://tools.ietf.org/html/rfc7231#section-6.4.7) for more info.

### exclude

- Default: `[]`

An array of routes patterns for which redirection should be disabled.

## Using with [Nuxt.js](https://github.com/nuxt/nuxt.js)

Add the `redirect-ssl` to the [`serverMiddleware`](https://nuxtjs.org/api/configuration-servermiddleware#usage) array within in the [nuxt.config.js](https://nuxtjs.org/api/configuration-server) file is the preferred usage:

```js
import redirectSSL from 'redirect-ssl'

export default {
  serverMiddleware: [
    redirectSSL.create({
      enabled: process.env.NODE_ENV === 'production'
     }),
  ]
}
```

You will still need to install this package within your project for it work.

## License

MIT. Made with 💖

<!-- Refs -->
[npm-v-src]: https://img.shields.io/npm/v/redirect-ssl?style=flat-square
[npm-v-href]: https://npmjs.com/package/redirect-ssl

[npm-d-src]: https://img.shields.io/npm/dm/redirect-ssl?style=flat-square
[npm-d-href]: https://npmjs.com/package/redirect-ssl

[ci-src]: https://img.shields.io/github/workflow/status/unjs/redirect-ssl/ci/master?style=flat-square
[ci-href]: https://github.com/unjs/redirect-ssl/actions?query=workflow%3Aci


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [3.0.0](https://github.com/unjs/redirect-ssl/compare/v2.0.1...v3.0.0) (2021-05-17)


### ⚠ BREAKING CHANGES

* **pkg:** add exports field
* update is-https to 4.x

### Features

* **pkg:** add exports field ([597ac39](https://github.com/unjs/redirect-ssl/commit/597ac395cac0dfccf9482539d738475cf9c9d1e3))
* update is-https to 4.x ([cbeab94](https://github.com/unjs/redirect-ssl/commit/cbeab94ab3d4d2f64fa36443bd8ec58a497eb6dd))


### Bug Fixes

* use ascii underscore char in variable name ([#39](https://github.com/unjs/redirect-ssl/issues/39)) ([e393ad6](https://github.com/unjs/redirect-ssl/commit/e393ad6b11fe76e43934f9737fdc5c1696e0806c))

### [2.0.1](https://github.com/unjs/redirect-ssl/compare/v2.0.0...v2.0.1) (2021-04-08)

## [2.0.0](https://github.com/unjs/redirect-ssl/compare/v1.4.1...v2.0.0) (2020-05-31)


### ⚠ BREAKING CHANGES

* xForwardedProto renamed to trustProxy
* redirect changed to enabled and default value is always true (see example in docs)
* there might be behaviour changes for express-like frameworks

### Features

* `enabled` option ([f06e398](https://github.com/unjs/redirect-ssl/commit/f06e3982a61bd3e6a98fdc09ab4aab85575b0e70))
* improve redirectHost type ([88cafc0](https://github.com/unjs/redirect-ssl/commit/88cafc0fb54c6507286d5dfa9c30a5368003513b))
* rewrite to typescript ([d8460a7](https://github.com/unjs/redirect-ssl/commit/d8460a73cd29328e055203181900bf2fa4d011b1))
* trustProxy option ([1c8cb3d](https://github.com/unjs/redirect-ssl/commit/1c8cb3d8c769a961f96f442e922242030c0d6645))
* write redirectURL to response too ([763165b](https://github.com/unjs/redirect-ssl/commit/763165b221a7b37957618bfe7f7f7b533790fa96))


### Bug Fixes

* fix types and redirectURL with non standard port ([e728797](https://github.com/unjs/redirect-ssl/commit/e72879708dc4a664edd3366997f87965d6d99dfc))


### fat

* update is-https to 2.x ([d4267e7](https://github.com/unjs/redirect-ssl/commit/d4267e76a2e176ed792a03f80b3ffebf7e39ea7e))

<a name="1.4.1"></a>
## [1.4.1](https://github.com/nuxt-community/redirect-ssl/compare/v1.4.0...v1.4.1) (2019-08-08)


### Bug Fixes

* remove space from URL when port is not 443 ([#16](https://github.com/nuxt-community/redirect-ssl/issues/16)) ([a73ca54](https://github.com/nuxt-community/redirect-ssl/commit/a73ca54))



<a name="1.4.0"></a>
# [1.4.0](https://github.com/nuxt-community/redirect-ssl/compare/v1.3.0...v1.4.0) (2018-10-10)


### Features

* add exclude options which adds the ability to disable SSL redirect on some routes ([#5](https://github.com/nuxt-community/redirect-ssl/issues/5)) ([83867e2](https://github.com/nuxt-community/redirect-ssl/commit/83867e2))


test/server.ts
import https from 'https'
import connect from 'connect'
import pem from 'pem'
import redirectSSL from '../src'

const HOST = 'localhost'
const HTTP_PORT = 8080
const HTTPS_PORT = 8443

const app = connect()

app.use(redirectSSL.create({
  statusCode: 302,
  redirectPort: HTTPS_PORT,
  exclude: [
    '^/health$'
  ]
}))

app.use('/', (_, res) => {
  res.end('Hi!')
})

pem.createCertificate({ days: 1, selfSigned: true }, (err, keys) => {
  https.createServer({ key: keys.serviceKey, cert: keys.certificate }, app).listen(HTTPS_PORT)
  if (err) {
    console.error(err)
  }
  console.log(`Listining on https://${HOST}:${HTTPS_PORT}`)
})

app.listen(HTTP_PORT, () => {
  console.log(`Listining on http://${HOST}:${HTTP_PORT}`)
  console.log(`No SSL redirect on excluded URL: http://${HOST}:${HTTP_PORT}/health`)
})


src/index.ts
import { IncomingMessage, ServerResponse } from 'http'
import isHTTPS from 'is-https'

// Default options
const defaults = {
  trustProxy: true,
  redirectPort: 443,
  redirectHost: '',
  redirectUnknown: true,
  enabled: true,
  statusCode: 307,
  exclude: []
}
type Options = Partial<typeof defaults>

interface Middleware {
  (req: IncomingMessage, res: ServerResponse, next?: Function): void
  create: (options: Options) => Middleware
}

const isExcluded = (url: string, patterns = []) => patterns.some(pattern => url.match(pattern))

// Creates new middleware using provided options
function create (_options?: Options): Middleware {
  const options = { ...defaults, ..._options }

  const _port = options.redirectPort === 443 ? '' : (':' + options.redirectPort)

  function redirectSSL (req: IncomingMessage, res: ServerResponse, next?: Function) {
    const url = req.url || ''

    if (!options.enabled || isExcluded(url, options.exclude)) {
      return next && next()
    }

    const _isHttps = isHTTPS(req, options.trustProxy)
    const shouldRedirect = options.redirectUnknown ? !_isHttps : _isHttps === false

    if (shouldRedirect) {
      const _redirectHost = (options.redirectHost || req.headers.host || '').split(':')[0]
      const _redirectURL = 'https://' + _redirectHost + _port + url
      res.writeHead(options.statusCode, { Location: _redirectURL })
      return res.end(_redirectURL)
    }

    return next && next()
  }

  redirectSSL.create = create

  return redirectSSL
}

// Export a new instance using defaults
export default create({})


index.js
const isHTTPS = require('is-https')

// Default options
const defaults = {
  xForwardedProto: true,
  redirectPort: 443,
  redirectHost: undefined,
  redirectUnknown: true,
  statusCode: 307,
  redirect: process.env.NODE_ENV === 'production',
  exclude: []
}

const isExcluded = function (url, patterns = []) {
  return patterns.some(pattern => url.match(pattern))
}

// Creates new middleware using provided options
function create (options) {
  const {
    xForwardedProto,
    redirectPort,
    redirectHost,
    statusCode,
    redirectUnknown,
    redirect,
    exclude
  } = Object.assign({}, defaults, options)
  const _port = redirectPort === 443 ? '' : (':' + redirectPort)

  return function redirectSSL (req, res, next) {
    if (redirect && !isExcluded(req.url, exclude)) {
      const _isHttps = isHTTPS(req, xForwardedProto)
      const shouldRedirect = _isHttps === false || (redirectUnknown && _isHttps === null)
      if (shouldRedirect) {
        const ـredirectURL = 'https://' + (redirectHost || req.headers.host) + _port + req.url
        res.writeHead(statusCode, { Location: ـredirectURL })
        return res.end()
      }
    }

    return next()
  }
}

// Create a new instance using defaults
const instance = create({})

// Assign create to instance
instance.create = create

// Export default instance
module.exports = instance
